packages/@react-spectrum/s2/src/CoachMark.tsx

```diff
@@ -10,417 +10,192 @@
-});
-
-const actionButtonSize = {
-  XS: 'XS',
-  S: 'XS',
-  M: 'S',
-  L: 'M',
-  XL: 'L'
-} as const;
-
-export const CoachMarkContext = createContext<ContextValue<CoachMarkProps, HTMLElement>>({});
-
-export const CoachMark = forwardRef((props: CoachMarkProps, ref: ForwardedRef<HTMLElement>) => {
-  let colorScheme = useContext(ColorSchemeContext);
+export const CoachMark = forwardRef((props: CoachMarkProps, ref: ForwardedRef<HTMLDivElement>) => {
   [props, ref] = useContextProps(props, ref, CoachMarkContext);
-  let {UNSAFE_style} = props;
-  let {size = 'M'} = props;
+  let {UNSAFE_style, UNSAFE_className = ''} = props;
   let popoverRef = useObjectRef(ref);
-
-  let children = (
-    <Provider
-      values={[
-        [ImageContext, {alt: '', styles: image}],
-        [TextContext, {
-          slots: {
-            [DEFAULT_SLOT]: {},
-            title: {styles: title({size})},
-            description: {styles: description({size})},
-            steps: {styles: steps}
-          }
-        }],
-        [KeyboardContext, {styles: keyboard}],
-        [ContentContext, {styles: content({size})}],
-        [DividerContext, {size: 'S'}],
-        [FooterContext, {styles: footer}],
-        [ActionMenuContext, {
-          isQuiet: true,
-          size: actionButtonSize[size],
-          // @ts-ignore
-          'data-slot': 'menu',
-          styles: actionMenu
-        }]
-      ]}>
-      <ImageCoordinator>
-        {props.children}
-      </ImageCoordinator>
-    </Provider>
-  );
+  let state = useContext(OverlayTriggerStateContext);
+  let {triggerRef} = useContext(InternalCoachMarkContext);
+  let fallbackTriggerRef = useObjectRef(useRef<HTMLElement>(null));
+  triggerRef = triggerRef ?? fallbackTriggerRef;
+
+  let {overlayProps, placement} = useOverlayPosition({
+    targetRef: triggerRef,
+    overlayRef: popoverRef,
+    placement: props.placement,
+    offset: 16,
+    crossOffset: -18 // made up
+  });
+
+  let prevOpen = useRef(false);
+  useLayoutEffect(() => {
+    if (state?.isOpen && !prevOpen.current) {
+      popoverRef.current?.showPopover();
+      internalContainer.current?.showPopover();
+    } else if (!state?.isOpen && prevOpen.current) {
+      popoverRef.current?.hidePopover();
+      internalContainer.current?.hidePopover();
+    }
+    prevOpen.current = state?.isOpen ?? false;
+  }, [state?.isOpen]);
+
+  let {keyboardProps} = useKeyboard({
+    onKeyDown: (e) => {
+      if (e.key === 'Escape') {
+        state?.close();
+        return;
+      }
+      e.continuePropagation();
+    }
+  });
+  // Have to put the portal in a div with popover="manual" so it is also in the top layer and renders on top of the coachmark
+  // Note, this isn't great because I'm not honestly sure what would happen if the page scrolled or their is a parent which affects that placement, is
+  // top-layer unaffected by parent stacking context? scroll parent?
+  let internalContainer = useRef<HTMLDivElement>(null);
```

Comment:
could remove this if we say menus aren't allowed in coachmarks, would need more examples and somewhere to put a 'X' button maybe?